### DISCUSSION TITLE
Save all changes feature

### DISCUSSION THREAD

**Discussion by zewa666** [ghiscoding/slickgrid-universal#1454] [Ideas]
Hey @ghiscoding.

finally I've started our company project with a dynamic OData based backend that provides column and grid configs for dynamically rendering and handling CRUD with a Slickgrid.

so far we're tracking changes in an editQueue, much like some examples you've made, and send updates/adds with a save all button to our backend, refreshing the current backend query on success. deletes are currently handled as one-offs but I plan to look into the rowselection option to also support bulk deletes.

our next step is to now improve the save all button to check all modified rows, iterate over their cells and trigger the validators to see if all client side validations are covered before executing the backend call. invalid cells will be highlighted with a individual cell css class applied.

so far the ideas. now I'm sure we're not the first ones to think about this and I recall our discussions to generalize the editQueue concept. now I wonder whether we should ideally think bigger and create a plugin to take care of all the parts involved.

would you see something like that as too big of a concern for slickgrid-universal itself and better kept in user land?

**Comment by ghiscoding** [comment]
do you mean keeping track of the visual styling of what's saved and what not (similar to Example 12)? I mean this goes back to our previous discussion that it could be nice to move that to some form of Edit Queue Service of some kind. I don't personally plan to work on that but if you wish to start then I would totally accept a PR for a new feature (it could be built-in or as an optional plugin if you think it could be huge).  Because yes we tend to copy over and over the same kind of code to do this so it would make sense to move this to a centralized service in the future. 

The only thing I'm not sure about in what you said is for validation of the queue?

**Reply by zewa666** [reply]
right so similar to example 12 I'd highlight cells that are modified and invalid due to a failing validator. if any cell is invalid the save feature should be blocked. hence to check that I'd iterate over all cells of the editQueue to see if thats the case

also plugin wise this could perhaps be part of the backendService to be used with both Graphql or OData backends. that would also remove the need for yet an additional plugin. so users could define a "save" method callback on the backend service the service than would have the editQueue and could e.g prompt whether changes should be discarded or saved before pagination etc

**Reply by ghiscoding** [reply]
but typically when the cell validator fails, it won't let you continue so I'm not sure how that would fit with your approach? For example below, it checks a regular expression of "Task x" where x is a number, if I remove the text "Task" it won't let me save (or continue in the case of a queueing). though I think I'm missing the event handler to display the error in the case shown below.

![msedge_81eZFE9UsP](https://github.com/ghiscoding/slickgrid-universal/assets/643976/90750223-e072-4038-b982-9969802d73b4)

**Reply by ghiscoding** [reply]
> also plugin wise this could perhaps be part of the backendService to be used with both Graphql or OData backends. that would also remove the need for yet an additional plugin. so users could define a "save" method callback on the backend service the service than would have the editQueue and could e.g prompt whether changes should be discarded or saved before pagination etc

well I guess we could do that, I can think of 2 possible approaches. 1- add a Validation to BackendService API or 2- some kind of event cancellation (like what we do with `onBeforeEditCell`, if the event calls `stopPropagation()` it won't go further)

**Reply by zewa666** [reply]
ohh ok I missed that point, so validations are intercepting the editors save call if they fail. that means in the case of copy/pasting multiple cells the failed validations would just ignore the pasted values and keep the cells which succeed?

**Reply by ghiscoding** [reply]
hmm I'm not sure about that one, you would have to give it a try, I almost never use copy+paste in grids myself, so...

**Reply by zewa666** [reply]
ok so the behavior there is that an editor is created on the fly to apply the new value. if the validation fails it applies the default of the respective inputEditor.

what it doesnt do is trigger an onValidationError. I'll look into creating a PR for that

**Reply by ghiscoding** [reply]
BTW when I made the Tooltip PR last week, I noticed that you made a mistake when you wrote the Cypress test, you complained that Cypress was wrong and you had to use the `force: true` on the button click but in reality you forgot to click Edit before clicking Update (perhaps the page was refreshed at that time, dunno), so Cypress was correct all along hahahaha 🤣  Anyway, I forgot to mention it to you before, so now you know 😋 below is the code I used to correct the mistake and add tooltip tests. 

![image](https://github.com/ghiscoding/slickgrid-universal/assets/643976/fcb24144-1232-4f02-8c2c-f987b0c8afe0)

**Reply by zewa666** [reply]
are you implying I made a mistake? well if its just a single one I'd say not bad 😅

thanks for pointing out the issue, glad to see its not due tooling.

**Reply by ghiscoding** [reply]
I was just having a laugh, don't take it too seriously 😄

**Reply by zewa666** [reply]
all good ;)
